ee/clickhouse/views/person.py

```diff
@@ -1,30 +1,93 @@
-from typing import Callable, Dict, Optional, Tuple
+import json
+from typing import Any, Callable, Dict, List, Optional, Tuple, cast
 
+from django.db.models.query import Prefetch, QuerySet
 from rest_framework.decorators import action
 from rest_framework.exceptions import NotFound
 from rest_framework.request import Request
 from rest_framework.response import Response
 
 from ee.clickhouse.client import sync_execute
 from ee.clickhouse.models.person import delete_person
+from ee.clickhouse.models.property import parse_prop_clauses
 from ee.clickhouse.queries.clickhouse_retention import ClickhouseRetention
 from ee.clickhouse.queries.clickhouse_stickiness import ClickhouseStickiness
 from ee.clickhouse.queries.funnels import ClickhouseFunnelPersons, ClickhouseFunnelTrendsPersons
 from ee.clickhouse.queries.trends.lifecycle import ClickhouseLifecycle
 from ee.clickhouse.sql.person import GET_PERSON_PROPERTIES_COUNT
 from posthog.api.person import PersonViewSet
-from posthog.api.utils import format_next_url, format_offset_absolute_url
-from posthog.constants import INSIGHT_FUNNELS, FunnelVizType
+from posthog.api.utils import format_offset_absolute_url
+from posthog.constants import FunnelVizType
 from posthog.decorators import cached_function
+from posthog.exceptions import UnsupportedFeature
 from posthog.models import Event, Filter, Person
 
+BASE_FILTER_PERSONS_QUERY = """
+SELECT person.id FROM person
+LEFT OUTER JOIN (
+    SELECT * FROM person_distinct_id WHERE person_distinct_id.team_id = %(team_id)s
+) person_distinct_id
+ON person.id = person_distinct_id.person_id
+WHERE team_id = %(team_id)s"""
+
+
+def filter_persons_ch(team_id: int, request: Request, queryset: QuerySet) -> QuerySet:
+    # Keep functionality in sync with posthog/api/person.py
+    params: Dict[str, Any] = {"team_id": team_id}
+    and_conditions: List[str] = []
+    if request.GET.get("id"):
+        raise UnsupportedFeature("filtering persons by field `id`")
+    if request.GET.get("uuid"):
+        params["uuids"] = request.GET["uuid"].split(",")
+        and_conditions.append("AND has(%(uuids)s, toString(person.id))")
+    if request.GET.get("search"):
+        parts = request.GET["search"].split(" ")
+        contains = []
+        for part_index, part in enumerate(parts):
+            if ":" in part:
+                matcher, key = part.split(":")
+                if matcher == "has":
+                    # Matches for example has:email or has:name
+                    params[f"has_key_{part_index}"] = key
+                    and_conditions.append(f"AND JSONHas(person.properties, %(has_key_{part_index})s)")
+            else:
+                contains.append(part)
+        params["icontains"] = f'%{" ".join(contains)}%'
+        and_conditions.append(
+            f"""
+        AND (
+            person.properties ILIKE %(icontains)s
```

Comment:
This is really really sketchy. 

1. We should only match on certain parsed properties not an opaque json blob - the edge cases introduced are kind of insane.
2. This won't be able to make use of materialized columns either
3. This won't match postgres' behavior

Reply:
You know, I agree! But this _does_ match Postgres's behavior, which is the point. `filter_persons()`, which this is an analogue of, literally does `properties__icontains` – and this is reflected in the SQL by `UPPER(posthog_person.properties::string) LIKE UPPER(test_string)` (`UPPER` instead of `ILIKE` I presume because the former is [a _bit_ faster in Postgres](https://www.peterbe.com/plog/upper-vs-ilike).

Reply:
Ah! I didn't check out the postgres implementation - I assumed it was based off of email property. :sweat_smile: 

Do the two behave exactly identically though?

I do think this is worth visiting from a product standpoint though before committing to this on clickhouse.

Reply:
They are indeed exactly identical, and actually this is _better_ in ClickHouse, because `properties` is already a UTF-8 string in ClickHouse, while the Postgres version had to (and still does) cast JSONB to a string, which is not free.
I would argue to keep this as is, simply because this is so ridiculously cheap compute-wise, while covering a case that would be really difficult to handle otherwise: search among keys and values of _nested_ properties. I'm okay with this being slightly quirky as a tradeoff.